src/Grpc.Net.Client/Internal/GrpcEventSource.cs

```diff
@@ -0,0 +1,144 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.Diagnostics.Tracing;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Grpc.Core;
+
+namespace Grpc.Net.Client.Internal
+{
+    internal class GrpcEventSource : EventSource
+    {
+        public static readonly GrpcEventSource Log = new GrpcEventSource();
+
+        private PollingCounter? _totalCallsCounter;
+        private PollingCounter? _currentCallsCounter;
+        private PollingCounter? _messagesSentCounter;
+        private PollingCounter? _messagesReceivedCounter;
+        private PollingCounter? _callsFailedCounter;
+        private PollingCounter? _callsDeadlineExceededCounter;
+
+        private long _totalCalls;
+        private long _currentCalls;
+        private long _messageSent;
+        private long _messageReceived;
+        private long _callsFailed;
+        private long _callsDeadlineExceeded;
+
+        internal GrpcEventSource()
+            : base("Grpc.Net.Client")
+        {
+        }
+
+        // Used for testing
+        internal GrpcEventSource(string eventSourceName)
+            : base(eventSourceName)
+        {
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Is this important?

Reply:
I believe it is required because WriteEvent needs to capture metadata off the method attribute.

I copied them from https://github.com/aspnet/AspNetCore/blob/04bf1bf32e13d35d4d7f61f9cbeb9bfa22e0c617/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L41

Reply:
That doesn't sound right, seems like that could easily break code if the compiler could remove attributes by optimizing a method.

I couldn't find any docs about this being required, and SignalR doesn't have it and our counters seem to be working...